src/zjs_web_sockets.c

```diff
@@ -0,0 +1,802 @@
+    DBG_PRINT("connection made, context %p error %d\n", context, error);
+    ws_connection_t *new = zjs_malloc(sizeof(ws_connection_t));
+    memset(new, 0, sizeof(ws_connection_t));
+    new->tcp_sock = context;
+    new->rbuf = zjs_malloc(DEFAULT_WS_BUFFER_SIZE);
+    memset(new->rbuf, 0, sizeof(DEFAULT_WS_BUFFER_SIZE));
+    new->rptr = new->wptr = new->rbuf;
+    new->server = handle->server;
+    new->trigger_msg_id = zjs_add_c_callback(new, trigger_data);
+    if (jerry_value_is_function(handle->accept_handler)) {
+        new->accept_handler_id = zjs_add_callback(handle->accept_handler,
+                                                  new->server_handle->server,
+                                                  new,
+                                                  post_accept_handler);
+    } else {
+        new->accept_handler_id = -1;
+    }
+    new->next = connections;
+    connections = new;
+
+    ret = net_context_recv(context, tcp_received, 0, new);
+    if (ret < 0) {
+        ERR_PRINT("Cannot receive TCP packet (family %d)\n",
+            net_context_get_family(context));
+        // this seems to mean the remote exists but the connection was not made
+        zjs_trigger_event(handle->server, "error", NULL, 0, NULL, NULL);
+        return;
+    }
+}
+
+static ZJS_DECL_FUNC(ws_server)
+{
+    ZJS_VALIDATE_ARGS_OPTCOUNT(optcount, Z_OBJECT, Z_OPTIONAL Z_FUNCTION);
+
+    int ret;
+    double port = 0;
+    bool backlog = 0;
+    double max_payload = 0;
+
+    server_handle_t *handle = zjs_malloc(sizeof(server_handle_t));
+    if (!handle) {
+        return zjs_error("out of memory");
+    }
+    memset(handle, 0, sizeof(server_handle_t));
+
+    jerry_value_t server = jerry_create_object();
+
+    zjs_obj_get_double(argv[0], "port", &port);
+    zjs_obj_get_boolean(argv[0], "backlog", &backlog);
+    zjs_obj_get_boolean(argv[0], "clientTracking", &handle->track);
+    zjs_obj_get_double(argv[0], "maxPayload", &max_payload);
+    jerry_value_t accept_handler = zjs_get_property(argv[0], "acceptHandler");
+    if (jerry_value_is_function(accept_handler)) {
+        handle->accept_handler = jerry_acquire_value(accept_handler);
+    }
+
+    CHECK(net_context_get(AF_INET6, SOCK_STREAM, IPPROTO_TCP,
+                &handle->tcp_sock));
+
+    struct sockaddr_in6 my_addr6 = { 0 };
+
+    my_addr6.sin6_family = AF_INET6;
+    my_addr6.sin6_port = htons((int)port);
+
+    CHECK(net_context_bind(handle->tcp_sock,
+                           (struct sockaddr *)&my_addr6,
+                           sizeof(struct sockaddr_in6)));
+    CHECK(net_context_listen(handle->tcp_sock, (int)backlog));
+    CHECK(net_context_accept(handle->tcp_sock, tcp_accepted, 0, handle));
+
+    if (optcount) {
+        zjs_add_event_listener(handle->server, "listening", argv[1]);
+        zjs_trigger_event(handle->server, "listening", NULL, 0, NULL, NULL);
+    }
+
+    handle->server = server;
+    handle->max_payload = (uint16_t)max_payload;
```

Comment:
This handle is the server handle. When I set the server objects native handle I register a free function (free_server()), line 790.